cmd/controller-manager/app/controllermanager.go

```diff
@@ -46,22 +50,35 @@ var knownControllers = map[string]startControllerFunc{
 
 var controllersDisabledByDefault = sets.New(MonitorControllerName)
 
-// startControllerFunc is responsible for constructing and starting a controller. startControllerFunc should be
-// asynchronous and an error is only returned if we fail to start the controller.
-type startControllerFunc func(ctx context.Context, controllerCtx *controllercontext.Context) error
-
 // Run starts the controller manager according to the given options.
 func Run(ctx context.Context, opts *options.Options) {
 	controllerCtx, err := createControllerContext(opts)
 	if err != nil {
 		klog.Fatalf("Error creating controller context: %v", err)
 	}
 
+	if opts.EnableProfiling {
+		go func() {
+			if err := http.ListenAndServe("0.0.0.0:6060", nil); err != nil {
+				klog.Errorf("Failed to start pprof server: %v", err)
+			}
+		}()
+	}
+
+	handler := healthcheck.NewMutableHealthCheckHandler()
+	handler.AddLivezChecker("ping", healthz.Ping)
+
+	var healthzAdaptor *leaderelection.HealthzAdaptor
+	if opts.EnableLeaderElect {
```

Comment:
Can this block be combined with the block below that also checks for `opts.EnableLeaderElect`?

Reply:
if we combine it with the block below, we may start the healthcheck server before the leaderelection check is installed.  will that be undesirable?

Reply:
Why don't we run healthcheck as the last step?

Reply:
`elector.Run` is run as the last step as it blocks until either `ctx` is done or the lease is lost, allowing `main` to return once the leader lease is lost.

we can still run healthcheck as the last step if you wish, but we would have to configure the `OnStoppedLeading` callback to cancel `ctx` in order to be notified when we lose the leader lease.

Reply:
In that case, can we 1. create leader elector; 2. register livez; 3. run elector?